tests/agent/python/flaskapp/smoke.py

```diff
@@ -0,0 +1,258 @@
+import argparse
+import time
+from tornado import ioloop, httpclient
+from elasticsearch import Elasticsearch
+from datetime import datetime, timedelta
+import subprocess
+import logging
+import json
+import copy
+import uuid
+
+p_uid = uuid.uuid4().hex
+
+num_reqs = 0    # use with care
+
+logger = logging.getLogger("logger")
+logger.setLevel(logging.INFO)
+handler = logging.StreamHandler()
+handler.setLevel(logging.INFO)
+formatter = logging.Formatter('[%(asctime)s] [%(process)s] [%(levelname)s] [%(funcName)s - %(lineno)d]  %(message)s')
+handler.setFormatter(formatter)
+logger.propagate = False
+logger.addHandler(handler)
+
+es = Elasticsearch(['localhost:9200'])
+
+
+def handle(r):
+    global num_reqs
+    try:
+        assert r.code == 200
+        num_reqs -= 1
+        if num_reqs == 0:
+            logger.info("Stopping tornado I/O loop")
+            ioloop.IOLoop.instance().stop()
+
+    except AssertionError:
+        num_reqs == 0
+        ioloop.IOLoop.instance().stop()
+        logger.error("Bad response, aborting: {} - {} ({})".format(r.code, r.error, r.request_time))
+
+
+def run_process(lang):
+    if lang == 'node':
+        cmd = ['node', '../../nodejs/app.js']
+    elif lang == 'python':
+        cmd = ['gunicorn', '-w', '4', '-b', '0.0.0.0:8081', 'app:app']
+    p = subprocess.Popen(cmd)
+    return p
+
+
+def send_batch(nreqs):
+    global num_reqs
+    global p_uid
+
+    http_client = httpclient.AsyncHTTPClient(max_clients=4)
+    for _ in range(nreqs):
+        num_reqs += 1
+        endpoint = "foo" if num_reqs % 2 == 0 else "bar"
+        url = "http://localhost:8081/" + endpoint + '?' + p_uid
+        http_client.fetch(url, handle, method='GET', connect_timeout=90, request_timeout=120)
+
+    logger.info("Starting tornado I/O loop")
+    ioloop.IOLoop.instance().start()
+
+
+class Worker:
+
+    def __init__(self, lang):
+        self.l = lang
+
+    def __enter__(self):
+        self.p = run_process(self.l)
+        time.sleep(3)
+        logger.info("Process ready {}".format(self.p.pid))
+
+    def __exit__(self, *args):
+        logger.info("Flushing queues")
+        time.sleep(2)
+        self.p.terminate()
+        time.sleep(10)
```

Comment:
What exactly is it that needs to happen in the time you sleep at `enter` and `exit` ? Could we use something like `wait_for_xyz` to reduce the waiting time, or adapt it based on how many requests are made?

Reply:
processes (eg gunicorn) need to start and be ready to accept requests, node agent needs time to flush the queue (that is the `2`), python agent also needs time to flush the queue upon SIGTERM (thats the `10`), then apm-server / libbeat needs time to pipe events to ES...

i don't like the `wait-for-log` approach for this, because then any kind of error looks like a timeout error and you don't know what is going on underneath, which is not great if you are executing this manually and watching the output. 
OTHO, if tests fail consistently under a reasonable waiting time, you then know something is up.

I don't worry much about execution time, because i don't intend to run it very often, at least for now...

Reply:
Thank you for the explanation. I was worried about the time to sleep being set according to the number of requests you used to make. But it sounds like the time to wait is a fixed time based on agent handling. With this I am fine, since as you said the tests are not intended to be included in the fully automated test runs.